### PR TITLE
BFD-2845: RDS/IAM authentication

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[BFD-2845](https://jira.cms.gov/browse/BFD-2845)

**User Story or Bug Summary:**

As an Aurora/RDS security best-practice, we should use IAM to authenticate database access. This will resolve several long-standing issues related to enabling and managing user and service account access and password rotations.

---

### What Does This PR Do?

Adds support for using IAM based authentication for RDS Aurora.  This is an optional feature controlled by a single new configuration setting, `DATABASE_AUTH_TYPE` (ETL pipeline and db migrator) or `bfdServer.db.authType` (API server), which can have one of two possible values:

- `JDBC`: Default value.  Uses simple authentication with a single password.
- `RDS`: Uses IAM authentication with a temporary token instead of a password.

Token lookup is implemented using a subclass of `HikariDataSource` with an overridden `getPassword()` method that performs the lookup at most once per 5 minutes.  A new `DataSourceFactory` interface was added that knows which type of data source to create based on the authentication type in use.

As this is the fourth AWS service being added to the code base, this PR also consolidates the parsing of common AWS options that previously had different configuration settings for each service.  

| New Setting | Old Settings | Description |
|---|---|---|
| `AWS_REGION` | `SSM_REGION` <br /> `DB_MIGRATOR_SQS_REGION` | AWS region.  Defaults to us-east-1. |
| `AWS_ENDPOINT` |`SSM_ENDPOINT` <br /> `S3_ENDPOINT_URI` <br /> `DB_MIGRATOR_SQS_ENDPOINT` | Endpoint override URI used with localstack. |
| `AWS_ACCESS_KEY` | `SSM_ACCESS_KEY` <br /> `S3_ACCESS_KEY` <br /> `DB_MIGRATOR_SQS_ACCESS_KEY` | Access key used with localstack. |
| `AWS_SECRET_KEY` | `SSM_SECRET_KEY` <br /> `S3_SECRET_KEY` <br /> `DB_MIGRATOR_SQS_SECRET_KEY` | Secret key used with localstack. |

#### Reader Notes

Most of the changes in this PR are either:

- Switching from passing around `DatabaseOptions` to passing around `DataSourceFactory`.
- Using the new env variables names for the common AWS settings.
- Using `AwsClientConfig` as the source of AWS common settings.

#### Configuration Notes

Configuring the pipeline and migrator:

- Set `DATABASE_AUTH_TYPE` environment variable to `RDS`
- Set `DATABASE_USERNAME` to the IAM user id.
- Set `DATABASE_PASSWORD` to any string - it will be ignored but cannot be left blank or missing.
- Ensure that `DATABASE_URL` has both a host and port defined.

Configuring the FHIR server:

- Set the new property `bfdServer.db.authType` to `RDS`.
- Set the property `bfdServer.db.username` to the IAM user id.
- Set the property `bfdServer.db.password` to anything (or leave it blank).
- Ensure that the property `bfdServer.db.url` has both a host and port defined.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [x] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [x] Yes
      * [ ] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

The two yes answers are technically yes but are very modest.

The new dependency is just using another AWS SDK API jar.

```
        <dependency>
            <!-- Used to interact with AWS' RDS service -->
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>rds</artifactId>
        </dependency>
```

The "modification of our security controls" is simply that we are using IAM based authentication to access our database instead of a fixed password.

https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html}

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/DirectRdaLoadApp.java [hunk 1]
```diff
@@ -158,4 +177,19 @@ private static RdaLoadOptions readRdaLoadOptionsFromProperties(ConfigLoader opti
...
         jobConfig.build(), grpcConfig, new RdaServerJob.Config(), 0, idHasherConfig);
   }
+
+  /**
+   * Reads aws client settings from the configuration.
+   *
+   * @param options the config options to use
+   * @return the aws config settings
+   */
+  @Nullable
```
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/DirectRdaLoadApp.java:187]: stared at this for a bit, and can't quite connect the dots....
rdsConfig init'd to null; but I don't see where it actually gets a value?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/LoadRdaJsonApp.java [hunk 2]
```diff
@@ -193,6 +207,10 @@ private static class Config {
     private Config(ConfigLoader options) {
       hashPepper = options.hexBytes("hash.pepper");
       hashIterations = options.intValue("hash.iterations");
+      dbAuthType =
+          options
+              .enumOption("database.authType", DatabaseOptions.AuthenticationType.class)
+              .orElse(DatabaseOptions.AuthenticationType.JDBC);
       dbUrl = options.stringValue("database.url", "jdbc:hsqldb:mem:LoadRdaJsonApp");
```
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/LoadRdaJsonApp.java:214]: So we are stiil supporting HSQL?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/RdaPipelineTestUtils.java [hunk 3]
```diff
@@ -91,10 +93,17 @@ public static void runTestWithTemporaryDb(Class<?> testClass, Clock clock, Datab
     // the HSQLDB database will be destroyed when this connection is closed
```
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/RdaPipelineTestUtils.java:93]: ditto on HSQL?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/DirectRdaLoadApp.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/DirectRdaLoadApp.java:187]: lol, will fix.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/LoadRdaJsonApp.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc-apps/src/main/java/gov/cms/bfd/pipeline/rda/grpc/apps/LoadRdaJsonApp.java:214]: That's default that was already there.  There are still some places in the code base that reference hsqldb (including this).  I believe there is a ticket in the backlog to remove them.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/RdaPipelineTestUtils.java (hunk 3 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/RdaPipelineTestUtils.java:93]: Ditto on these remaining references will be removed in a future ticket.

**Review by mackec1 — DISMISSED** [review]
Looks good!

**Review by StewGoin — APPROVED** [review]
Please and thank you :)

**Review by mackec1 — APPROVED** [review]

**Review by StewGoin — APPROVED** [review]
Approved to proceed from my perspective.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java [hunk 4]
```diff
@@ -471,6 +463,17 @@ static ConfigLoader createConfigLoader(Function<String, String> getenv) {
...
   }
 
+  /**
+   * Build a {@link ConfigLoader} for use in tests that just uses default values and those returned
+   * by the provided function.
+   *
+   * @param getenv function used to access environment variables (provided explicitly for testing)
+   * @return appropriately configured {@link ConfigLoader}
+   */
+  static ConfigLoader createConfigLoaderForTesting(Function<String, String> getenv) {
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java:473]: Don't super love adding test-only code in app code if we can avoid it. Could we just call the builder with these values in the test where we need to set it up? (Or a shared test util if its used a lot?)

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/AppConfigurationTest.java [hunk 5]
```diff
@@ -362,7 +426,7 @@ public void testLoadRdaSourceConfig() {
   @Test
   public void testLoadRdaServerJobConfig() {
     var settingsMap = new HashMap<String, String>();
-    var configLoader = ConfigLoader.builder().addSingle(settingsMap::get).build();
+    final var configLoader = AppConfiguration.createConfigLoaderForTesting(settingsMap::get);
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/AppConfigurationTest.java:429]: Does this need to be loaded for each test, or could we set it up once for all tests?

**Review by lsmitchell — COMMENTED** [review]
Looks good! A couple small nonblocking naggles.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java (hunk 4 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java:473]: We could and I considered doing that.  But then the provide default values map would need to be made public.  I felt that this revealed less private information and, honestly, this is only test only in the sense that it's only used by them.  Nothing about it depends on any test libraries or statics that are used only during testing.

I could just as easily have named it `createMinimalConfigLoader()`.

File: apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/AppConfigurationTest.java (hunk 5 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/test/java/gov/cms/bfd/pipeline/app/AppConfigurationTest.java:429]: Every test uses its own config settings.

**Review by lsmitchell — APPROVED** [review]

**Pull request merged**